### PR TITLE
Fixes #18774 - Can provide content overrides to AK in bulk

### PR DESCRIPTION
**Pull request by parthaa**
Api bindings to provide content overrides in bulk
for activation keys

### CODE REVIEW

**Comment by waldenraines** [comment]
Works for me but will let someone else have a peak at the code.

**Comment by ehelms** [comment]
Seems like we should deprecate the singular case in favor of the array case so we end up with a single way to do it.

**Comment by parthaa** [comment]
> Seems like we should deprecate the singular case in favor of the array case so we end up with a single way to do it.

@ehelms addressed.

File: app/controllers/katello/api/v2/activation_keys_controller.rb [hunk 1]
```diff
@@ -197,13 +197,28 @@ def remove_subscriptions
...
+    end
+    param :content_overrides, Array, :desc => N_("Array of Content override parameters") do
       param :content_label, String, :desc => N_("Label of the content"), :required => true
       param :value, [0, 1, "default"], :desc => N_("Override to 0/1, or 'default'"), :required => true
     end
     def content_override
-      content_override = validate_content_overrides(params[:content_override])
-      @activation_key.set_content_override(content_override[:content_label], 'enabled', content_override[:value])
+      if params[:content_override]
+        ::Foreman::Deprecation.api_deprecation_warning("The parameter content_override will be removed in Katello 3.4. Please update to use the uploads parameter.")
```
Comment (ehelms) [inline app/controllers/katello/api/v2/activation_keys_controller.rb]: I think you mean '3.5' given 3.4 is the next release.

File: app/controllers/katello/api/v2/activation_keys_controller.rb [hunk 2]
```diff
@@ -197,13 +197,28 @@ def remove_subscriptions
...
+      end
+      content_overrides = params[:content_overrides] ? params[:content_overrides] : [params[:content_override]]
+      content_overrides = [] unless content_overrides
+      content_override_values = content_overrides.map do |content_override_params|
+        validate_content_overrides(content_override_params)
+      end
+
+      content_override_values.each do |content_override|
+        @activation_key.set_content_override(content_override[:content_label], 'enabled', content_override[:value])
```
Comment (ehelms) [inline app/controllers/katello/api/v2/activation_keys_controller.rb]: Do we have to make a call to Candlepin for each provided override or can they be sent in bulk?

**Comment by parthaa** [comment]
> What is the performance of setting multiple content overrides? Currently, just doing one has always been entirely too slow.

@ehelms  updated. Now it tries to bulk up multiple content overides in one call.

File: app/models/katello/content_override.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
+module Katello
```
Comment (ehelms) [inline app/models/katello/content_override.rb]: This should probably go to `app/lib` since its not a Rails model.

Comment (ehelms) [inline app/models/katello/content_override.rb]: Also this looks like a good file for some clean new unit tests.

File: app/models/katello/content_override.rb [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+      @value = value
+    end
+
+    def to_hash
+      ret = {"content_label" => @content_label, "name" => @name}
+      ret["value"] = @value if @value
+      ret
+    end
+
+    def self.create_enabled(content_params)
```
Comment (ehelms) [inline app/models/katello/content_override.rb]: Why opt for this class method to create an object when we Ruby has the `new` method which can create and return an object that can be used?

File: app/models/katello/content_override.rb (hunk 3 above)
Comment (parthaa) [inline app/models/katello/content_override.rb]: hmm https://github.com/Katello/katello/blob/master/app/models/katello/pool_with_quantities.rb indicates that we do have models here that are not rails related

File: app/models/katello/content_override.rb (hunk 4 above)
Comment (parthaa) [inline app/models/katello/content_override.rb]: Let me put this in perspective, ContentOverride accepts a content_label, name/key to override and value. In Katello so far we have only used one particular value for  'name' be it AK/Content Hosts and that name or key is "enabled".

More over I need a central place to hold the logic  that says parse the content_params being send and create me an "enabled" ContentOverride object from it.

ContentOverride.new(label,{value: 1}) will not tell me that the name value here should be "enabled" . So a better approach is use the new method to set generic content override where you will pass me label, name and value that are not specifically related to setting "enabled" overriding attributes.

Comment (jlsherrill) [inline app/models/katello/content_override.rb]: right, i guess i'd argue it'd make more sense to keep this 'limitation' at the api level.  Since our api only supports 'enabled', it'd call:

ContentOverride.new(content_label, 'enabled', value)

Comment (parthaa) [inline app/models/katello/content_override.rb]: Look at  the 2 snippets below
https://github.com/Katello/katello/pull/6667/files#diff-beaa1cde8d4964406e1f688ae9de5befL164

https://github.com/Katello/katello/pull/6667/files#diff-65d2a2cbc038a622df645394a941f518L227

Both controllers have duplicated logic. I felt Model will be the right place to reuse this part and create the correct one.

**Comment by parthaa** [comment]
@ehelms @jlsherrill I made the following changes based on your suggestions

1.  Added a new concern "ContentOverridesController" with a method that just handles the validation + creation of content override part. This gets included in the ActivationKeysController ( the hosts controller in the subsequent PR)
1.  Removed the create_enabled off the model

Hopefully this satisfies both of your concerns.

**Comment by parthaa** [comment]
> Also this looks like a good file for some clean new unit tests.

Also added that.

File: app/models/katello/content_override.rb (hunk 3 above)
Comment (ehelms) [inline app/models/katello/content_override.rb]: That doesn't mean that that file is correct either :)

File: app/controllers/katello/concerns/api/v2/content_overrides_controller.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+        override = ::Katello::ContentOverride.new(content_params[:content_label])
+        case content_params[:value].to_s.downcase
+        when 'default'
+          override.enabled = nil
+        when '1', "yes"
+          override.enabled = '1'
+        when '0', "no"
+          override.enabled = '0'
+        end
```
Comment (ehelms) [inline app/controllers/katello/concerns/api/v2/content_overrides_controller.rb]: Why did you move this logic out of the ContentOverride class? Given how specific it is to a content override, this seemed like good information to store there and retrieve from the instance of the class by passing in the `content_params[:value]`.

File: app/controllers/katello/concerns/api/v2/content_overrides_controller.rb [hunk 6]
```diff
@@ -0,0 +1,30 @@
+module Katello
+  module Concerns
+    module Api::V2::ContentOverridesController
+      extend ActiveSupport::Concern
+
+      # overriding_object => pass it either an activation key or content host
+      def validate_content_overrides_enabled(content_params, overriden_object)
+        valid_values = ["yes", "no", "0", "1", "default"]
```
Comment (ehelms) [inline app/controllers/katello/concerns/api/v2/content_overrides_controller.rb]: I could see these being a constant defined on `Katello::ContentOverride`

File: app/models/katello/content_override.rb (hunk 3 above)
Comment (jlsherrill) [inline app/models/katello/content_override.rb]: There isn't anything about 'models' that requires them to be tied to active record.  Reading some debate online it seems to really be up to each project to decide their own 'convention'.

Comment (ehelms) [inline app/models/katello/content_override.rb]: Let the bike shedding begin, I kid slightly. I feel like there are _more_ files that aren't active record in `app/lib` than in the models directory. If not `app/lib` then perhaps put them in a sub-directory of models to prevent confusion. I know I _should_ check the class inheritance when looking at a file inside the `models` directory, but given how convention over configuration based Rails is, my tendency is going to be to think anything in there is tied to AR.

File: app/controllers/katello/concerns/api/v2/content_overrides_controller.rb (hunk 5 above)
Comment (parthaa) [inline app/controllers/katello/concerns/api/v2/content_overrides_controller.rb]: I made that change after discussing with @jlsherrill . It was deemed that "http request params" should not be passed to model objects. I needed to reuse this validate method in 3 places so moved it to a generic controller.

File: app/models/katello/content_override.rb (hunk 3 above)
Comment (ehelms) [inline app/models/katello/content_override.rb]: I won't hold the PR up either way, just expressing my opinion on how this can lead to developers wasting their time (or it might just be me wasting my time :))

Comment (jlsherrill) [inline app/models/katello/content_override.rb]: yeah, i tend to not necessarily make that particular assumption (or at least discovering it very early in the rails console).  I'm not arguing against only making models hold AR classes, just that a) there is no global convention and b) we haven't set a convention for our project.   So i think if we want to make some convention we should document it and fix any existing classes.

File: app/controllers/katello/concerns/api/v2/content_overrides_controller.rb (hunk 6 above)
Comment (ehelms) [inline app/controllers/katello/concerns/api/v2/content_overrides_controller.rb]: The API does not specify yes and no as valid values. Where does the discrepancy come in?

File: app/controllers/katello/concerns/api/v2/content_overrides_controller.rb [hunk 7]
```diff
@@ -0,0 +1,30 @@
+module Katello
+  module Concerns
+    module Api::V2::ContentOverridesController
+      extend ActiveSupport::Concern
+
+      # overriding_object => pass it either an activation key or content host
```
Comment (ehelms) [inline app/controllers/katello/concerns/api/v2/content_overrides_controller.rb]: typo here -- the method defines this as overriden_object

File: app/models/katello/content_override.rb (hunk 3 above)
Comment (ehelms) [inline app/models/katello/content_override.rb]: :white_flag:

Comment (ehelms) [inline app/models/katello/content_override.rb]: :bike: :house:

File: app/controllers/katello/concerns/api/v2/content_overrides_controller.rb (hunk 6 above)
Comment (parthaa) [inline app/controllers/katello/concerns/api/v2/content_overrides_controller.rb]: I would argue the model only cares about ["1", "0" or nil] . its just a place holder. The yes/no/default these are all mainly a controller formatting thing.

File: app/controllers/katello/concerns/api/v2/content_overrides_controller.rb [hunk 8]
```diff
@@ -0,0 +1,30 @@
+module Katello
+  module Concerns
+    module Api::V2::ContentOverridesController
+      extend ActiveSupport::Concern
+
+      # overriden_object => pass it either an activation key or content host
+      def validate_content_overrides_enabled(content_params, overriden_object)
+        valid_values = ["yes", "no", "0", "1", "default"]
```
Comment (parthaa) [inline app/controllers/katello/concerns/api/v2/content_overrides_controller.rb]: @ehelms Dont know why your comments on these got removed.
All I was trying to say was this no different than saying "ForemanCast::to_bool" other than the fact there is a "default" also. Feels like something very specific to controller data param massaging. If want it to be a constant I can just declare it in the controller.

File: app/controllers/katello/concerns/api/v2/content_overrides_controller.rb (hunk 6 above)
Comment (ehelms) [inline app/controllers/katello/concerns/api/v2/content_overrides_controller.rb]: I am still confused, if you look at -(https://github.com/Katello/katello/pull/6656/files#diff-65d2a2cbc038a622df645394a941f518R202)  yes/no are not valid values to pass in.
